# **Procesamiento de Lenguaje Natural**

## Maestría en Inteligencia Artificial Aplicada
#### Tecnológico de Monterrey
#### Prof Luis Eduardo Falcón Morales

### **Actividad individual: Sistema inteligente para búsqueda de talento**

* **Nombre:** Santiago Espinosa Domínguez
* **Matrícula:** A01747478





# Sistema Inteligente para Búsqueda de Talento

## 1. Selección de la vacante real

Para esta actividad se seleccionó una vacante real relacionada con el área de Inteligencia Artificial, Ciencia de Datos y Machine Learning. La vacante elegida será utilizada como entrada principal del sistema inteligente, ya que sus requisitos, responsabilidades y competencias servirán como referencia para comparar los perfiles de los candidatos sintéticos.

El objetivo del sistema es analizar una base de conocimiento formada por currículums sintéticos en formato PDF y PNG, extraer información relevante de cada perfil y recomendar los cinco candidatos más adecuados para la vacante seleccionada.

### Vacante seleccionada

**Empresa:** Microsoft  

**Portal consultado:** Microsoft Careers

**Nota:** La información de la vacante fue tomada como referencia de una publicación real relacionada con perfiles de Data Scientist / Machine Learning Engineer.

**Nombre de la vacante:** Data Scientist / Applied Scientist / Machine Learning Engineer  

**Área:** Inteligencia Artificial, Ciencia de Datos y Machine Learning  

### Justificación de la elección

Se seleccionó una vacante relacionada con Ciencia de Datos y Machine Learning porque permite evaluar de forma clara las competencias técnicas de los candidatos, como programación en Python, manejo de datos, aprendizaje automático, estadística, análisis de información y uso de herramientas de inteligencia artificial.

Además, este tipo de vacante es adecuada para construir un sistema inteligente de búsqueda de talento, ya que los requisitos técnicos pueden compararse semánticamente contra los currículums de los candidatos. Esto permite identificar qué perfiles tienen mayor coincidencia con la vacante, no solo por palabras exactas, sino también por similitud de experiencia, habilidades y proyectos.

### Descripción del puesto

La vacante está orientada a un perfil capaz de analizar grandes volúmenes de datos, construir modelos de Machine Learning, desarrollar soluciones basadas en Inteligencia Artificial y colaborar con equipos multidisciplinarios para resolver problemas de negocio mediante el uso de datos. La persona seleccionada deberá participar en el diseño, entrenamiento, evaluación y despliegue de modelos predictivos, así como en la interpretación de resultados para apoyar la toma de decisiones dentro de la organización.

### Competencias requeridas

- Pensamiento analítico.
- Resolución de problemas.
- Comunicación efectiva de resultados.
- Trabajo en equipo.
- Capacidad para traducir problemas de negocio en soluciones basadas en datos.
- Aprendizaje continuo.
- Razonamiento estadístico.
- Capacidad para documentar modelos y experimentos.

### Responsabilidades

- Analizar conjuntos de datos estructurados y no estructurados.
- Construir modelos de Machine Learning para resolver problemas específicos.
- Evaluar el desempeño de modelos mediante métricas adecuadas.
- Colaborar con equipos de ingeniería, producto y negocio.
- Generar reportes e interpretaciones claras de los resultados.
- Participar en el desarrollo de soluciones de Inteligencia Artificial.
- Mantener buenas prácticas de documentación, experimentación y reproducibilidad.

### Requisitos técnicos

- Programación en Python.
- Conocimiento de Machine Learning supervisado y no supervisado.
- Manejo de librerías como pandas, scikit-learn, NumPy o PySpark.
- Conocimiento de SQL.
- Experiencia con análisis de datos.
- Conocimiento de estadística aplicada.
- Deseable experiencia con procesamiento de lenguaje natural.
- Deseable conocimiento de plataformas en la nube como Azure, AWS o Google Cloud.
- Deseable experiencia con bases de datos vectoriales, embeddings o sistemas de recuperación de información.

In [3]:
# Información estructurada de la vacante seleccionada
vacante = {
    "empresa": "Microsoft",
    "nombre_vacante": "Data Scientist / Machine Learning Engineer",
    "descripcion": """
    Perfil orientado al análisis de datos, construcción de modelos de Machine Learning,
    desarrollo de soluciones de Inteligencia Artificial y colaboración con equipos
    multidisciplinarios para resolver problemas de negocio mediante datos.
    """,
    "competencias": [
        "pensamiento analítico",
        "resolución de problemas",
        "comunicación efectiva",
        "trabajo en equipo",
        "aprendizaje continuo",
        "razonamiento estadístico"
    ],
    "responsabilidades": [
        "analizar conjuntos de datos",
        "construir modelos de Machine Learning",
        "evaluar modelos predictivos",
        "interpretar resultados",
        "colaborar con equipos técnicos y de negocio",
        "documentar experimentos"
    ],
    "requisitos_tecnicos": [
        "Python",
        "Machine Learning",
        "SQL",
        "pandas",
        "scikit-learn",
        "PySpark",
        "estadística",
        "NLP",
        "cloud computing",
        "embeddings",
        "bases de datos vectoriales"
    ]
}

## 2. Generación de CVs sintéticos

Para esta actividad se generaron 20 currículums sintéticos utilizando inteligencia artificial. Debido a que la actividad busca simular un sistema de búsqueda de talento sin comprometer información personal real, los CVs no corresponden a personas reales.

Cada CV incluye información profesional relevante para el análisis de talento, como resumen profesional, formación académica, experiencia laboral, habilidades técnicas, habilidades blandas, certificaciones, idiomas y proyectos relevantes.

Los perfiles fueron generados con tres niveles de experiencia:

- **Junior:** perfiles con poca experiencia profesional, normalmente de 0 a 2 años. Suelen contar con conocimientos básicos, proyectos académicos o prácticas profesionales.
- **Intermedio o Semi-Senior:** perfiles con experiencia media, normalmente de 3 a 5 años. Cuentan con experiencia laboral comprobable, participación en proyectos reales y dominio de herramientas técnicas.
- **Senior:** perfiles con experiencia avanzada, normalmente de más de 6 años. Suelen liderar proyectos, diseñar arquitecturas, tomar decisiones técnicas y coordinar equipos.

## 3. Importancia de no incluir datos personales sensibles

En la generación de los CVs sintéticos se evitó incluir información como nombre completo, fotografía, género, edad, nacionalidad, estado civil, domicilio o cualquier otro dato personal sensible.

Esto es importante porque un sistema de búsqueda de talento debe enfocarse en las competencias, experiencia, formación, habilidades y proyectos de las personas candidatas, no en características personales que puedan generar sesgos o discriminación.

Excluir este tipo de información ayuda a que el sistema sea más justo, ético y objetivo. Además, permite proteger la privacidad de las personas y reducir el riesgo de que el modelo favorezca o descarte candidatos por razones no relacionadas con su capacidad profesional.

## 4. Prompt utilizado para generar los CVs sintéticos

El siguiente prompt fue utilizado para generar los currículums sintéticos mediante un modelo de lenguaje:

**Prompt:**

Genera 20 currículums sintéticos en español para una base de talento orientada a una vacante de Data Scientist / Machine Learning Engineer. Los CVs no deben incluir nombre real, fotografía, género, edad, nacionalidad, estado civil, dirección ni datos personales sensibles.

Cada CV debe incluir las siguientes secciones:

- Resumen profesional.
- Formación académica.
- Experiencia laboral.
- Habilidades técnicas.
- Habilidades blandas.
- Certificaciones.
- Idiomas.
- Proyectos relevantes.

Genera perfiles variados con los siguientes niveles de experiencia:

- 5 perfiles Junior, con 0 a 2 años de experiencia.
- 10 perfiles Intermedio o Semi-Senior, con 3 a 5 años de experiencia.
- 5 perfiles Senior, con más de 6 años de experiencia.

Incluye perfiles cercanos a la vacante y algunos menos cercanos para poder evaluar el desempeño del sistema. Los tipos de perfiles deben incluir:

- Científico de datos.
- Ingeniero de datos.
- Analista Business Intelligence.
- Machine Learning Engineer.
- Desarrollador de software.
- Arquitecto de software o perfil híbrido.

Cada CV debe tener un identificador anónimo, por ejemplo CV_01, CV_02, CV_03, etc.

## 5. Tipos de perfiles generados

Para construir una base de conocimiento variada, se generaron perfiles relacionados directamente con la vacante y otros perfiles cercanos, pero no necesariamente ideales. Esto permite evaluar si el sistema inteligente es capaz de jerarquizar correctamente a los mejores candidatos.

La distribución de perfiles fue la siguiente:

| Tipo de perfil | Cantidad | Relación con la vacante |
|---|---:|---|
| Científico de datos | 4 | Alta |
| Machine Learning Engineer | 4 | Alta |
| Ingeniero de datos | 4 | Media-alta |
| Analista Business Intelligence | 3 | Media |
| Desarrollador de software | 3 | Media-baja |
| Arquitecto de software / perfil híbrido | 2 | Media |

Esta variedad permite validar si la búsqueda semántica prioriza perfiles con experiencia en Machine Learning, Python, análisis de datos, estadística, NLP, modelos predictivos y herramientas de inteligencia artificial.

In [4]:
# Base de CVs sintéticos

cvs_sinteticos = [
    {
        "id": "CV_01",
        "nivel": "Junior",
        "perfil": "Científico de datos",
        "contenido": """
        Resumen profesional:
        Perfil junior en ciencia de datos con experiencia académica en análisis exploratorio,
        modelos predictivos básicos y visualización de datos.

        Formación académica:
        Licenciatura en Ingeniería en Sistemas Computacionales.

        Experiencia laboral:
        Practicante de análisis de datos durante 1 año, apoyando en limpieza de datos,
        reportes y construcción de modelos básicos de clasificación.

        Habilidades técnicas:
        Python, pandas, NumPy, scikit-learn, SQL, visualización con matplotlib.

        Habilidades blandas:
        Aprendizaje rápido, comunicación, organización, trabajo en equipo.

        Certificaciones:
        Curso de fundamentos de Machine Learning.
        Curso de análisis de datos con Python.

        Idiomas:
        Español nativo, inglés intermedio.

        Proyectos relevantes:
        Modelo de clasificación para predecir abandono de clientes usando scikit-learn.
        Dashboard exploratorio de ventas con Python.
        """
    },
    {
        "id": "CV_02",
        "nivel": "Junior",
        "perfil": "Machine Learning Engineer",
        "contenido": """
        Resumen profesional:
        Perfil junior orientado a Machine Learning con experiencia en proyectos académicos
        de clasificación, regresión y procesamiento básico de lenguaje natural.

        Formación académica:
        Licenciatura en Ciencia de Datos.

        Experiencia laboral:
        Becario de Machine Learning durante 1 año, participando en preparación de datos
        y pruebas de modelos supervisados.

        Habilidades técnicas:
        Python, scikit-learn, pandas, SQL, Git, Jupyter Notebook, NLP básico.

        Habilidades blandas:
        Pensamiento analítico, curiosidad, colaboración, adaptabilidad.

        Certificaciones:
        Certificación introductoria en Inteligencia Artificial.
        Curso de Python para ciencia de datos.

        Idiomas:
        Español nativo, inglés intermedio-avanzado.

        Proyectos relevantes:
        Clasificador de sentimientos en reseñas de productos.
        Modelo de regresión para estimación de precios.
        """
    },
    {
        "id": "CV_03",
        "nivel": "Junior",
        "perfil": "Analista Business Intelligence",
        "contenido": """
        Resumen profesional:
        Perfil junior en Business Intelligence con experiencia en creación de reportes,
        dashboards y análisis descriptivo de indicadores de negocio.

        Formación académica:
        Licenciatura en Administración con enfoque en analítica.

        Experiencia laboral:
        Analista BI trainee durante 1 año, desarrollando tableros y reportes periódicos.

        Habilidades técnicas:
        SQL, Power BI, Excel avanzado, Python básico, análisis descriptivo.

        Habilidades blandas:
        Comunicación, atención al detalle, orientación a resultados, responsabilidad.

        Certificaciones:
        Curso de Power BI.
        Curso de SQL básico.

        Idiomas:
        Español nativo, inglés intermedio.

        Proyectos relevantes:
        Dashboard de ventas mensuales.
        Reporte automatizado de indicadores operativos.
        """
    },
    {
        "id": "CV_04",
        "nivel": "Junior",
        "perfil": "Ingeniero de datos",
        "contenido": """
        Resumen profesional:
        Perfil junior en ingeniería de datos con conocimientos en bases de datos,
        procesamiento ETL y programación en Python.

        Formación académica:
        Ingeniería en Tecnologías de la Información.

        Experiencia laboral:
        Practicante de datos durante 2 años apoyando en extracción, transformación
        y carga de datos.

        Habilidades técnicas:
        Python, SQL, bases de datos relacionales, ETL básico, Spark básico.

        Habilidades blandas:
        Orden, solución de problemas, trabajo colaborativo, aprendizaje continuo.

        Certificaciones:
        Fundamentos de bases de datos.
        Introducción a Big Data.

        Idiomas:
        Español nativo, inglés intermedio.

        Proyectos relevantes:
        Pipeline ETL para integrar archivos CSV a una base SQL.
        Limpieza de datos de clientes para análisis posterior.
        """
    },
    {
        "id": "CV_05",
        "nivel": "Junior",
        "perfil": "Desarrollador de software",
        "contenido": """
        Resumen profesional:
        Desarrollador junior con experiencia en backend, APIs y bases de datos.
        Interés en evolucionar hacia soluciones de datos e inteligencia artificial.

        Formación académica:
        Ingeniería en Desarrollo de Software.

        Experiencia laboral:
        Desarrollador backend junior durante 2 años.

        Habilidades técnicas:
        Python, JavaScript, Flask, SQL, Git, APIs REST.

        Habilidades blandas:
        Trabajo en equipo, comunicación, pensamiento lógico, responsabilidad.

        Certificaciones:
        Curso de desarrollo backend con Python.
        Curso de Git y control de versiones.

        Idiomas:
        Español nativo, inglés intermedio.

        Proyectos relevantes:
        API para gestión de usuarios.
        Sistema web para registro de inventario.
        """
    }
]

In [5]:
# Se agregan 10 CVs sintéticos de nivel Intermedio / Semi-Senior

cvs_sinteticos.extend([
    {
        "id": "CV_06",
        "nivel": "Intermedio",
        "perfil": "Científico de datos",
        "contenido": """
        Resumen profesional:
        Científico de datos con 4 años de experiencia en modelos predictivos,
        análisis estadístico y comunicación de resultados para áreas de negocio.

        Formación académica:
        Maestría en Ciencia de Datos.

        Experiencia laboral:
        Científico de datos en empresa de retail, desarrollando modelos de segmentación,
        predicción de demanda y análisis de comportamiento de clientes.

        Habilidades técnicas:
        Python, pandas, NumPy, scikit-learn, SQL, estadística, visualización de datos,
        experimentación y validación de modelos.

        Habilidades blandas:
        Pensamiento analítico, comunicación ejecutiva, trabajo en equipo,
        orientación a resultados.

        Certificaciones:
        Machine Learning con Python.
        Estadística aplicada para ciencia de datos.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        Modelo de predicción de demanda semanal.
        Segmentación de clientes mediante clustering.
        Análisis de comportamiento de compra.
        """
    },
    {
        "id": "CV_07",
        "nivel": "Intermedio",
        "perfil": "Machine Learning Engineer",
        "contenido": """
        Resumen profesional:
        Machine Learning Engineer con 5 años de experiencia desarrollando,
        evaluando y desplegando modelos de aprendizaje automático en ambientes productivos.

        Formación académica:
        Ingeniería en Computación.

        Experiencia laboral:
        Ingeniero de Machine Learning en empresa tecnológica, trabajando en pipelines
        de entrenamiento, evaluación de modelos y despliegue de APIs predictivas.

        Habilidades técnicas:
        Python, scikit-learn, TensorFlow, PyTorch, MLflow, Docker, FastAPI,
        SQL, Git, modelos supervisados y no supervisados.

        Habilidades blandas:
        Resolución de problemas, colaboración, documentación técnica,
        pensamiento crítico.

        Certificaciones:
        Machine Learning Engineering.
        Docker para aplicaciones de datos.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        API para predicción de abandono de clientes.
        Pipeline de entrenamiento automatizado.
        Sistema de recomendación basado en comportamiento de usuarios.
        """
    },
    {
        "id": "CV_08",
        "nivel": "Intermedio",
        "perfil": "Ingeniero de datos",
        "contenido": """
        Resumen profesional:
        Ingeniero de datos con 4 años de experiencia en construcción de pipelines,
        integración de fuentes de datos y procesamiento distribuido.

        Formación académica:
        Ingeniería en Sistemas Computacionales.

        Experiencia laboral:
        Ingeniero de datos en empresa financiera, desarrollando procesos ETL,
        modelos de datos y flujos de procesamiento para analítica avanzada.

        Habilidades técnicas:
        Python, SQL, PySpark, Spark, Airflow, bases de datos relacionales,
        almacenamiento en la nube, ETL y modelado de datos.

        Habilidades blandas:
        Organización, pensamiento estructurado, colaboración, atención al detalle.

        Certificaciones:
        Big Data Fundamentals.
        Data Engineering with Spark.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        Pipeline de datos transaccionales con PySpark.
        Automatización de procesos ETL diarios.
        Integración de datos desde múltiples fuentes.
        """
    },
    {
        "id": "CV_09",
        "nivel": "Intermedio",
        "perfil": "Analista Business Intelligence",
        "contenido": """
        Resumen profesional:
        Analista Business Intelligence con 3 años de experiencia en análisis de datos,
        tableros ejecutivos y generación de indicadores para toma de decisiones.

        Formación académica:
        Licenciatura en Actuaría.

        Experiencia laboral:
        Analista BI en empresa de servicios, creando dashboards, reportes automatizados
        y análisis de tendencias para distintas áreas de negocio.

        Habilidades técnicas:
        SQL, Power BI, Tableau, Excel avanzado, Python, análisis descriptivo,
        modelado dimensional.

        Habilidades blandas:
        Comunicación, pensamiento analítico, atención al detalle, enfoque al negocio.

        Certificaciones:
        Power BI Data Analyst.
        SQL para análisis de datos.

        Idiomas:
        Español nativo, inglés intermedio-avanzado.

        Proyectos relevantes:
        Dashboard de indicadores financieros.
        Reporte de desempeño comercial.
        Automatización de consultas SQL para reportes semanales.
        """
    },
    {
        "id": "CV_10",
        "nivel": "Intermedio",
        "perfil": "Desarrollador de software",
        "contenido": """
        Resumen profesional:
        Desarrollador de software con 4 años de experiencia en backend, APIs,
        bases de datos y servicios web. Cuenta con conocimientos básicos de analítica.

        Formación académica:
        Ingeniería en Software.

        Experiencia laboral:
        Desarrollador backend en empresa tecnológica, construyendo servicios REST,
        integraciones con bases de datos y módulos para sistemas internos.

        Habilidades técnicas:
        Python, Java, Flask, FastAPI, SQL, PostgreSQL, Git, Docker,
        arquitectura básica de servicios.

        Habilidades blandas:
        Trabajo en equipo, solución de problemas, comunicación técnica,
        responsabilidad.

        Certificaciones:
        Desarrollo backend con Python.
        Fundamentos de arquitectura de software.

        Idiomas:
        Español nativo, inglés intermedio.

        Proyectos relevantes:
        API para gestión de inventarios.
        Sistema de autenticación para aplicación interna.
        Integración de servicios con bases de datos SQL.
        """
    },
    {
        "id": "CV_11",
        "nivel": "Intermedio",
        "perfil": "Machine Learning Engineer",
        "contenido": """
        Resumen profesional:
        Perfil semi-senior especializado en Machine Learning y NLP, con experiencia
        en clasificación de texto, embeddings y evaluación de modelos.

        Formación académica:
        Maestría en Inteligencia Artificial Aplicada.

        Experiencia laboral:
        Machine Learning Engineer en área de análisis de texto, desarrollando modelos
        para clasificación de documentos, extracción de información y búsqueda semántica.

        Habilidades técnicas:
        Python, scikit-learn, Transformers, embeddings, NLP, FAISS, SQL,
        evaluación de modelos, APIs de modelos de lenguaje.

        Habilidades blandas:
        Pensamiento crítico, comunicación, autonomía, documentación.

        Certificaciones:
        Natural Language Processing Specialization.
        Machine Learning avanzado.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        Clasificador de documentos legales.
        Motor de búsqueda semántica con embeddings.
        Sistema de extracción automática de entidades.
        """
    },
    {
        "id": "CV_12",
        "nivel": "Intermedio",
        "perfil": "Científico de datos",
        "contenido": """
        Resumen profesional:
        Científico de datos con experiencia en analítica predictiva, modelos de regresión,
        clasificación y análisis de datos para optimización de procesos.

        Formación académica:
        Ingeniería Industrial con especialidad en analítica.

        Experiencia laboral:
        Científico de datos en empresa logística, desarrollando modelos para estimación
        de tiempos, optimización de rutas y análisis de desempeño operativo.

        Habilidades técnicas:
        Python, pandas, scikit-learn, SQL, estadística, modelos predictivos,
        visualización de datos, experimentación.

        Habilidades blandas:
        Resolución de problemas, comunicación, pensamiento analítico,
        trabajo multidisciplinario.

        Certificaciones:
        Data Science Professional Certificate.
        Estadística para Machine Learning.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        Modelo de predicción de tiempos de entrega.
        Análisis de eficiencia operativa.
        Dashboard de desempeño logístico.
        """
    },
    {
        "id": "CV_13",
        "nivel": "Intermedio",
        "perfil": "Ingeniero de datos",
        "contenido": """
        Resumen profesional:
        Ingeniero de datos semi-senior con experiencia en procesamiento de grandes
        volúmenes de datos y construcción de arquitecturas para analítica.

        Formación académica:
        Ingeniería en Tecnologías Computacionales.

        Experiencia laboral:
        Ingeniero de datos en startup de tecnología financiera, implementando pipelines
        batch y procesos de integración de datos para modelos analíticos.

        Habilidades técnicas:
        Python, SQL, PySpark, Spark, Airflow, AWS, S3, Glue,
        bases de datos relacionales y no relacionales.

        Habilidades blandas:
        Planeación, colaboración, pensamiento estructurado, adaptabilidad.

        Certificaciones:
        AWS Cloud Practitioner.
        Spark para Big Data.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        Pipeline de scoring crediticio.
        Procesamiento de datos financieros con Spark.
        Arquitectura de datos para reportes analíticos.
        """
    },
    {
        "id": "CV_14",
        "nivel": "Intermedio",
        "perfil": "Arquitecto de software / perfil híbrido",
        "contenido": """
        Resumen profesional:
        Perfil híbrido con experiencia en arquitectura de software, backend y soluciones
        orientadas a datos. Ha colaborado con equipos de analítica e ingeniería.

        Formación académica:
        Ingeniería en Computación.

        Experiencia laboral:
        Arquitecto de soluciones junior en empresa tecnológica, diseñando componentes
        backend, integraciones y servicios para procesamiento de información.

        Habilidades técnicas:
        Python, Java, APIs REST, microservicios, SQL, Docker, arquitectura cloud,
        integración de datos.

        Habilidades blandas:
        Liderazgo técnico, comunicación, análisis de requerimientos, colaboración.

        Certificaciones:
        Fundamentos de arquitectura cloud.
        Diseño de microservicios.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        Diseño de arquitectura para plataforma de reportes.
        Integración de servicios de datos con APIs internas.
        Migración de monolito a microservicios.
        """
    },
    {
        "id": "CV_15",
        "nivel": "Intermedio",
        "perfil": "Analista Business Intelligence",
        "contenido": """
        Resumen profesional:
        Analista BI con 5 años de experiencia en inteligencia de negocios,
        indicadores estratégicos y análisis de datos comerciales.

        Formación académica:
        Licenciatura en Economía.

        Experiencia laboral:
        Analista BI en empresa de consumo, desarrollando tableros ejecutivos,
        análisis de ventas y reportes de comportamiento de clientes.

        Habilidades técnicas:
        SQL, Power BI, Tableau, Excel avanzado, Python básico,
        modelado de datos, análisis descriptivo.

        Habilidades blandas:
        Comunicación con negocio, atención al detalle, pensamiento analítico,
        orientación a resultados.

        Certificaciones:
        Business Intelligence Analyst.
        Tableau Fundamentals.

        Idiomas:
        Español nativo, inglés intermedio-avanzado.

        Proyectos relevantes:
        Dashboard nacional de ventas.
        Análisis de rentabilidad por segmento.
        Reporte mensual de KPIs comerciales.
        """
    }
])

In [7]:
# Se agregan 5 CVs sintéticos de nivel Senior

cvs_sinteticos.extend([
    {
        "id": "CV_16",
        "nivel": "Senior",
        "perfil": "Machine Learning Engineer",
        "contenido": """
        Resumen profesional:
        Machine Learning Engineer senior con 8 años de experiencia en diseño,
        entrenamiento, evaluación y despliegue de modelos de inteligencia artificial
        en ambientes productivos.

        Formación académica:
        Maestría en Inteligencia Artificial.

        Experiencia laboral:
        Líder técnico de Machine Learning en empresa tecnológica, coordinando equipos
        de ciencia de datos e ingeniería para construir soluciones predictivas,
        sistemas de recomendación y modelos de NLP.

        Habilidades técnicas:
        Python, TensorFlow, PyTorch, scikit-learn, MLflow, Docker, Kubernetes,
        FastAPI, SQL, Spark, NLP, embeddings, MLOps.

        Habilidades blandas:
        Liderazgo técnico, comunicación ejecutiva, mentoría, pensamiento estratégico,
        resolución de problemas complejos.

        Certificaciones:
        Machine Learning Engineering avanzado.
        MLOps Fundamentals.
        Arquitectura de soluciones en la nube.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        Plataforma de modelos predictivos para retención de clientes.
        Sistema de recomendación en producción.
        Arquitectura MLOps para monitoreo y versionamiento de modelos.
        """
    },
    {
        "id": "CV_17",
        "nivel": "Senior",
        "perfil": "Científico de datos",
        "contenido": """
        Resumen profesional:
        Científico de datos senior con 7 años de experiencia en analítica avanzada,
        modelado estadístico, experimentación y desarrollo de modelos predictivos
        para áreas de negocio.

        Formación académica:
        Doctorado en Estadística Aplicada.

        Experiencia laboral:
        Científico de datos senior en empresa financiera, liderando proyectos de
        scoring, segmentación, análisis de riesgo y modelos de predicción.

        Habilidades técnicas:
        Python, R, SQL, estadística avanzada, scikit-learn, modelos supervisados,
        modelos no supervisados, experimentación A/B, visualización de datos.

        Habilidades blandas:
        Pensamiento analítico, comunicación con stakeholders, liderazgo,
        toma de decisiones basada en datos.

        Certificaciones:
        Advanced Data Science.
        Statistical Learning.
        Experimentación y causalidad aplicada.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        Modelo de scoring de riesgo.
        Segmentación avanzada de clientes.
        Sistema de predicción de comportamiento financiero.
        """
    },
    {
        "id": "CV_18",
        "nivel": "Senior",
        "perfil": "Ingeniero de datos",
        "contenido": """
        Resumen profesional:
        Ingeniero de datos senior con 9 años de experiencia en arquitectura de datos,
        procesamiento distribuido, pipelines escalables y plataformas de Big Data.

        Formación académica:
        Maestría en Ingeniería de Software.

        Experiencia laboral:
        Arquitecto de datos en empresa global, diseñando plataformas de datos
        para analítica avanzada, inteligencia artificial y reportería ejecutiva.

        Habilidades técnicas:
        Python, SQL, PySpark, Spark, Airflow, Kafka, AWS, Databricks,
        bases de datos relacionales y NoSQL, modelado de datos, data lakes.

        Habilidades blandas:
        Liderazgo técnico, planeación, comunicación, pensamiento sistémico,
        gestión de equipos.

        Certificaciones:
        Data Engineering Professional.
        AWS Data Analytics.
        Databricks Lakehouse Fundamentals.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        Arquitectura de data lake empresarial.
        Pipeline de procesamiento masivo con Spark.
        Plataforma de datos para modelos de Machine Learning.
        """
    },
    {
        "id": "CV_19",
        "nivel": "Senior",
        "perfil": "Arquitecto de software / perfil híbrido",
        "contenido": """
        Resumen profesional:
        Arquitecto de software senior con 10 años de experiencia en diseño de sistemas,
        microservicios, plataformas cloud y soluciones tecnológicas orientadas a datos.

        Formación académica:
        Maestría en Arquitectura de Software.

        Experiencia laboral:
        Arquitecto de soluciones en empresa tecnológica, liderando el diseño de
        arquitecturas escalables para aplicaciones empresariales, integración de datos
        y servicios de inteligencia artificial.

        Habilidades técnicas:
        Python, Java, microservicios, APIs REST, Docker, Kubernetes, SQL,
        arquitectura cloud, AWS, Azure, diseño de sistemas distribuidos.

        Habilidades blandas:
        Liderazgo, negociación técnica, comunicación ejecutiva, visión estratégica,
        trabajo multidisciplinario.

        Certificaciones:
        Cloud Solutions Architect.
        Microservices Architecture.
        DevOps Foundations.

        Idiomas:
        Español nativo, inglés avanzado.

        Proyectos relevantes:
        Plataforma cloud para procesamiento de información.
        Arquitectura de microservicios para analítica empresarial.
        Integración de servicios de IA en sistemas internos.
        """
    },
    {
        "id": "CV_20",
        "nivel": "Senior",
        "perfil": "Desarrollador de software",
        "contenido": """
        Resumen profesional:
        Desarrollador de software senior con 8 años de experiencia en backend,
        APIs, sistemas empresariales y automatización de procesos. Cuenta con
        conocimientos complementarios en análisis de datos.

        Formación académica:
        Ingeniería en Sistemas Computacionales.

        Experiencia laboral:
        Desarrollador backend senior en empresa de servicios digitales, liderando
        el desarrollo de aplicaciones internas, APIs y servicios de integración.

        Habilidades técnicas:
        Python, Java, Flask, FastAPI, SQL, PostgreSQL, Docker, Git,
        arquitectura backend, pruebas unitarias, integración continua.

        Habilidades blandas:
        Liderazgo técnico, responsabilidad, comunicación, solución de problemas,
        mentoría a desarrolladores junior.

        Certificaciones:
        Backend Development Professional.
        Software Engineering Best Practices.
        Docker y despliegue de servicios.

        Idiomas:
        Español nativo, inglés intermedio-avanzado.

        Proyectos relevantes:
        API empresarial para gestión de procesos internos.
        Sistema backend para automatización de reportes.
        Plataforma de integración con servicios externos.
        """
    }
])

In [9]:
# Verificación final de cantidad de CVs sintéticos

print("Cantidad total de CVs sintéticos:", len(cvs_sinteticos))

Cantidad total de CVs sintéticos: 20


In [11]:
# Revisión de distribución por nivel de experiencia

import pandas as pd
df_cvs = pd.DataFrame(cvs_sinteticos)
df_cvs.groupby("nivel")["id"].count()

,id
nivel,
Intermedio,10
Junior,5
Senior,5


In [12]:
# Revisión de distribución por tipo de perfil

df_cvs.groupby("perfil")["id"].count()

,id
perfil,
Analista Business Intelligence,3
Arquitecto de software / perfil híbrido,2
Científico de datos,4
Desarrollador de software,3
Ingeniero de datos,4
Machine Learning Engineer,4


Con estos documentos se completó la base de 20 CVs sintéticos solicitada para la actividad. La base incluye 5 perfiles Junior, 10 perfiles Intermedio o Semi-Senior y 5 perfiles Senior.

También se incluyeron diferentes tipos de perfiles profesionales, como científicos de datos, ingenieros de datos, analistas BI, Machine Learning Engineers, desarrolladores de software y arquitectos de software o perfiles híbridos. Esta diversidad permite evaluar si el sistema inteligente logra identificar correctamente a los candidatos más cercanos a la vacante seleccionada.

## 6. Generación de archivos PDF y PNG de los CVs sintéticos

Para simular una base de conocimiento multimodal, los CVs sintéticos se guardaron en dos formatos diferentes: PDF y PNG. Esto permite que el sistema inteligente trabaje con documentos de texto y también con imágenes, como ocurriría en un proceso real de reclutamiento donde los CVs pueden recibirse en distintos formatos.

La distribución utilizada fue la siguiente:

- 1 CV Junior en formato PNG.
- 2 CVs Intermedio/Semi-Senior en formato PNG.
- 1 CV Senior en formato PNG.
- Los 16 CVs restantes en formato PDF.

Esta variedad permite probar la capacidad del sistema para procesar información proveniente de diferentes tipos de archivo.

In [13]:
# Instalación de librerías necesarias para crear archivos PDF y PNG

!pip install reportlab pillow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 8.1 MB/s eta 0:00:00


In [14]:
# Importación de librerías para generación de archivos

import os
from reportlab.lib.pagesizes import letter
from reportlab.pdfgen import canvas
from PIL import Image, ImageDraw, ImageFont
import textwrap

In [15]:
# Creación de carpeta para almacenar los CVs sintéticos

carpeta_cvs = "cvs_sinteticos"

os.makedirs(carpeta_cvs, exist_ok=True)

print("Carpeta creada:", carpeta_cvs)

Carpeta creada: cvs_sinteticos


In [16]:
# Función para guardar un CV en formato PDF

def guardar_cv_pdf(cv, carpeta):
    """
    Guarda el contenido de un CV sintético en formato PDF.
    """

    ruta_archivo = os.path.join(carpeta, f"{cv['id']}.pdf")

    c = canvas.Canvas(ruta_archivo, pagesize=letter)
    width, height = letter

    x = 50
    y = height - 50

    # Título del CV
    c.setFont("Helvetica-Bold", 14)
    c.drawString(x, y, f"{cv['id']} - {cv['nivel']} - {cv['perfil']}")

    y -= 30

    # Contenido del CV
    c.setFont("Helvetica", 10)

    lineas = cv["contenido"].split("\n")

    for linea in lineas:
        linea = linea.strip()

        if linea == "":
            y -= 10
            continue

        lineas_ajustadas = textwrap.wrap(linea, width=90)

        for linea_ajustada in lineas_ajustadas:
            if y < 50:
                c.showPage()
                c.setFont("Helvetica", 10)
                y = height - 50

            c.drawString(x, y, linea_ajustada)
            y -= 14

    c.save()

    return ruta_archivo

In [17]:
# Función para guardar un CV en formato PNG

def guardar_cv_png(cv, carpeta):
    """
    Guarda el contenido de un CV sintético en formato PNG.
    """

    ruta_archivo = os.path.join(carpeta, f"{cv['id']}.png")

    ancho = 1200
    alto = 1600

    imagen = Image.new("RGB", (ancho, alto), "white")
    draw = ImageDraw.Draw(imagen)

    try:
        fuente_titulo = ImageFont.truetype("DejaVuSans-Bold.ttf", 28)
        fuente_texto = ImageFont.truetype("DejaVuSans.ttf", 22)
    except:
        fuente_titulo = ImageFont.load_default()
        fuente_texto = ImageFont.load_default()

    x = 50
    y = 50

    # Título del CV
    draw.text(
        (x, y),
        f"{cv['id']} - {cv['nivel']} - {cv['perfil']}",
        fill="black",
        font=fuente_titulo
    )

    y += 50

    # Contenido del CV
    lineas = cv["contenido"].split("\n")

    for linea in lineas:
        linea = linea.strip()

        if linea == "":
            y += 15
            continue

        lineas_ajustadas = textwrap.wrap(linea, width=85)

        for linea_ajustada in lineas_ajustadas:
            draw.text(
                (x, y),
                linea_ajustada,
                fill="black",
                font=fuente_texto
            )
            y += 32

            if y > alto - 80:
                break

    imagen.save(ruta_archivo)

    return ruta_archivo

In [18]:
# Selección de CVs que se guardarán como PNG

cvs_png = ["CV_01", "CV_06", "CV_11", "CV_16"]
print("CVs que se guardarán como PNG:", cvs_png)

CVs que se guardarán como PNG: ['CV_01', 'CV_06', 'CV_11', 'CV_16']


In [19]:
# Generación de archivos PDF y PNG

archivos_generados = []
for cv in cvs_sinteticos:
    if cv["id"] in cvs_png:
        ruta = guardar_cv_png(cv, carpeta_cvs)
        formato = "PNG"
    else:
        ruta = guardar_cv_pdf(cv, carpeta_cvs)
        formato = "PDF"

    archivos_generados.append({
        "id": cv["id"],
        "nivel": cv["nivel"],
        "perfil": cv["perfil"],
        "formato": formato,
        "ruta": ruta
    })

df_archivos = pd.DataFrame(archivos_generados)

df_archivos

,id,nivel,perfil,formato,ruta
0,CV_01,Junior,Científico de datos,PNG,cvs_sinteticos/CV_01.png
1,CV_02,Junior,Machine Learning Engineer,PDF,cvs_sinteticos/CV_02.pdf
2,CV_03,Junior,Analista Business Intelligence,PDF,cvs_sinteticos/CV_03.pdf
3,CV_04,Junior,Ingeniero de datos,PDF,cvs_sinteticos/CV_04.pdf
4,CV_05,Junior,Desarrollador de software,PDF,cvs_sinteticos/CV_05.pdf
5,CV_06,Intermedio,Científico de datos,PNG,cvs_sinteticos/CV_06.png
6,CV_07,Intermedio,Machine Learning Engineer,PDF,cvs_sinteticos/CV_07.pdf
7,CV_08,Intermedio,Ingeniero de datos,PDF,cvs_sinteticos/CV_08.pdf
8,CV_09,Intermedio,Analista Business Intelligence,PDF,cvs_sinteticos/CV_09.pdf
9,CV_10,Intermedio,Desarrollador de software,PDF,cvs_sinteticos/CV_10.pdf


In [20]:
# Verificación de formatos generados

df_archivos.groupby(["nivel", "formato"])["id"].count()

nivel       formato
Intermedio  PDF        8
            PNG        2
Junior      PDF        4
            PNG        1
Senior      PDF        4
            PNG        1
Name: id, dtype: int64

La generación de archivos permitió construir una base de conocimiento multimodal compuesta por 20 CVs sintéticos. De estos documentos, 16 fueron guardados en formato PDF y 4 en formato PNG.

Los archivos PNG fueron distribuidos de acuerdo con las instrucciones de la actividad: un CV Junior, dos CVs Intermedio/Semi-Senior y un CV Senior. Esta distribución permite simular un escenario más realista, donde los CVs pueden recibirse en distintos formatos y el sistema debe ser capaz de procesarlos para recuperar información relevante.

## 7. Sistema inteligente multimodal para búsqueda de talento

En esta sección se construye un sistema inteligente para recuperar y recomendar candidatos a partir de una base de conocimiento formada por CVs sintéticos en formato PDF y PNG.

El sistema realiza las siguientes etapas:

1. Extracción de texto desde archivos PDF y PNG.
2. Construcción de una representación textual de la vacante.
3. Generación de embeddings para la vacante y para cada CV.
4. Construcción de una base vectorial.
5. Búsqueda semántica de los candidatos más similares a la vacante.
6. Selección y justificación de los 5 mejores candidatos.

Para los archivos PDF se extrae texto directamente del documento. Para los archivos PNG se utiliza OCR, ya que el texto se encuentra dentro de una imagen.

In [21]:
# Instalación de librerías necesarias

!apt-get install -y tesseract-ocr
!pip install pytesseract pypdf sentence-transformers faiss-cpu

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
tesseract-ocr is already the newest version (4.1.1-2.1build1).
0 upgraded, 0 newly installed, 0 to remove and 3 not upgraded.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 349.5/349.5 kB 2.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.5/18.5 MB 49.3 MB/s eta 0:00:00


In [22]:
# Importación de librerías

import os
import numpy as np
import pandas as pd
import pytesseract
import faiss

from PIL import Image
from pypdf import PdfReader
from sentence_transformers import SentenceTransformer

In [23]:
# Función para extraer texto de archivos PDF

def extraer_texto_pdf(ruta_archivo):
    texto = ""
    reader = PdfReader(ruta_archivo)
    for page in reader.pages:
        texto_pagina = page.extract_text()
        if texto_pagina:
            texto += texto_pagina + "\n"
    return texto

In [24]:
# Función para extraer texto de archivos PNG mediante OCR

def extraer_texto_png(ruta_archivo):
    imagen = Image.open(ruta_archivo)
    texto = pytesseract.image_to_string(imagen, lang="eng")
    return texto

In [25]:
# Extracción de texto de todos los CVs generados

textos_cvs = []

for _, row in df_archivos.iterrows():
    ruta = row["ruta"]
    formato = row["formato"]
    if formato == "PDF":
        texto = extraer_texto_pdf(ruta)
    elif formato == "PNG":
        texto = extraer_texto_png(ruta)
    else:
        texto = ""

    textos_cvs.append({
        "id": row["id"],
        "nivel": row["nivel"],
        "perfil": row["perfil"],
        "formato": formato,
        "ruta": ruta,
        "texto_extraido": texto
    })

df_textos_cvs = pd.DataFrame(textos_cvs)
df_textos_cvs.head()

,id,nivel,perfil,formato,ruta,texto_extraido
0,CV_01,Junior,Científico de datos,PNG,cvs_sinteticos/CV_01.png,CV.01- Junior-Cientafcode dates\n\nResumen pro...
1,CV_02,Junior,Machine Learning Engineer,PDF,cvs_sinteticos/CV_02.pdf,CV_02 - Junior - Machine Learning Engineer\nRe...
2,CV_03,Junior,Analista Business Intelligence,PDF,cvs_sinteticos/CV_03.pdf,CV_03 - Junior - Analista Business Intelligenc...
3,CV_04,Junior,Ingeniero de datos,PDF,cvs_sinteticos/CV_04.pdf,CV_04 - Junior - Ingeniero de datos\nResumen p...
4,CV_05,Junior,Desarrollador de software,PDF,cvs_sinteticos/CV_05.pdf,CV_05 - Junior - Desarrollador de software\nRe...


In [26]:
# Revisión rápida del texto extraído de un CV

print(df_textos_cvs.loc[0, "texto_extraido"][:1000])

CV.01- Junior-Cientafcode dates

Resumen profesional
Perfiljunioren cienciade datoscon experiencia acadsimica en andlisisexploratoria,

‘modelos predictive bascosyvisualizacian de datos

Formaciun acadmica

Licenciaturaen Ingenierty en SistemasComputacionales,

Experiencia laborat
Practcante de analissde datosdurante 1 ata apoyandoen limpieza de datos,

reportesy construccian de modelos besicosde clsificacin,

Habilidades taonicas

Python, pandas NumPy, scikitlearn SOL, visualizacian con matpotib

Habilidades blandas:

‘Aprendizaje rapida comunicacion organizacian,trabajoen equipo

Certfcaciones:
Cursode fundamentasde Machine Learning,

CCursode antlisisde datascon Python,

Idiomas:

Fspono nativa inglasintermedio.

Proyectos relevanios
Modetode clasificacin para predecir abandonode clientes usandoscikitlean.

Dashboard exploratoriode ventascon Python.



La extracción de texto permite convertir los archivos PDF y PNG en información procesable por el sistema. Esto es necesario porque los modelos de búsqueda semántica no trabajan directamente con documentos visuales, sino con representaciones textuales.

En el caso de los archivos PDF, el texto se extrae directamente. En el caso de los archivos PNG, se utiliza OCR para reconocer el contenido escrito dentro de la imagen.

In [28]:
# Construcción del texto de la vacante

texto_vacante = f"""
Empresa: {vacante['empresa']}

Nombre de la vacante: {vacante['nombre_vacante']}

Descripción:
{vacante['descripcion']}

Competencias requeridas:
{', '.join(vacante['competencias'])}

Responsabilidades:
{', '.join(vacante['responsabilidades'])}

Requisitos técnicos:
{', '.join(vacante['requisitos_tecnicos'])}
"""

print(texto_vacante)


Empresa: Microsoft

Nombre de la vacante: Data Scientist / Machine Learning Engineer

Descripción:

    Perfil orientado al análisis de datos, construcción de modelos de Machine Learning,
    desarrollo de soluciones de Inteligencia Artificial y colaboración con equipos
    multidisciplinarios para resolver problemas de negocio mediante datos.
    

Competencias requeridas:
pensamiento analítico, resolución de problemas, comunicación efectiva, trabajo en equipo, aprendizaje continuo, razonamiento estadístico

Responsabilidades:
analizar conjuntos de datos, construir modelos de Machine Learning, evaluar modelos predictivos, interpretar resultados, colaborar con equipos técnicos y de negocio, documentar experimentos

Requisitos técnicos:
Python, Machine Learning, SQL, pandas, scikit-learn, PySpark, estadística, NLP, cloud computing, embeddings, bases de datos vectoriales



In [29]:
# Carga del modelo de embeddings

modelo_embeddings = SentenceTransformer("sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2")

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:124: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/3.89k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/645 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/471M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/526 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.08M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [30]:
# Generación de embeddings para los CVs

lista_textos_cvs = df_textos_cvs["texto_extraido"].tolist()

embeddings_cvs = modelo_embeddings.encode(
    lista_textos_cvs,
    convert_to_numpy=True,
    normalize_embeddings=True
)
embeddings_cvs.shape

(20, 384)

In [31]:
# Generación del embedding de la vacante

embedding_vacante = modelo_embeddings.encode(
    [texto_vacante],
    convert_to_numpy=True,
    normalize_embeddings=True
)

embedding_vacante.shape

(1, 384)

In [32]:
# Construcción de la base de datos vectorial con FAISS

dimension = embeddings_cvs.shape[1]
index = faiss.IndexFlatIP(dimension)
index.add(embeddings_cvs)
print("Cantidad de CVs en la base vectorial:", index.ntotal)

Cantidad de CVs en la base vectorial: 20


Para construir la base vectorial se utilizó FAISS, una biblioteca especializada en búsqueda eficiente de vectores. Cada CV fue convertido en un embedding, es decir, una representación numérica de su contenido semántico.

También se generó un embedding para la vacante seleccionada. Posteriormente, la búsqueda se realiza comparando la similitud entre el vector de la vacante y los vectores de los CVs.

En este caso se utiliza similitud por producto interno sobre embeddings normalizados, lo cual equivale a una comparación basada en similitud coseno.

In [33]:
# Búsqueda semántica de los 5 mejores candidatos

top_k = 5
scores, indices = index.search(embedding_vacante, top_k)
resultados_recomendacion = []
for rank, idx in enumerate(indices[0]):
    candidato = df_textos_cvs.iloc[idx]
    resultados_recomendacion.append({
        "ranking": rank + 1,
        "id": candidato["id"],
        "nivel": candidato["nivel"],
        "perfil": candidato["perfil"],
        "formato": candidato["formato"],
        "score_similitud": float(scores[0][rank]),
        "ruta": candidato["ruta"],
        "texto_extraido": candidato["texto_extraido"]
    })

df_top5 = pd.DataFrame(resultados_recomendacion)
df_top5[["ranking", "id", "nivel", "perfil", "formato", "score_similitud"]]

,ranking,id,nivel,perfil,formato,score_similitud
0,1,CV_11,Intermedio,Machine Learning Engineer,PNG,0.677967
1,2,CV_16,Senior,Machine Learning Engineer,PNG,0.668851
2,3,CV_07,Intermedio,Machine Learning Engineer,PDF,0.658535
3,4,CV_02,Junior,Machine Learning Engineer,PDF,0.635409
4,5,CV_09,Intermedio,Analista Business Intelligence,PDF,0.583487


In [34]:
# Función para generar justificación básica de la recomendación

def generar_justificacion(texto_cv, perfil, nivel):
    texto_lower = texto_cv.lower()

    criterios = {
        "Python": "python" in texto_lower,
        "Machine Learning": "machine learning" in texto_lower or "aprendizaje automático" in texto_lower,
        "SQL": "sql" in texto_lower,
        "NLP": "nlp" in texto_lower or "procesamiento de lenguaje natural" in texto_lower,
        "PySpark/Spark": "pyspark" in texto_lower or "spark" in texto_lower,
        "Cloud": "aws" in texto_lower or "azure" in texto_lower or "cloud" in texto_lower,
        "Embeddings": "embedding" in texto_lower or "embeddings" in texto_lower,
        "MLOps": "mlops" in texto_lower,
        "Modelos predictivos": "modelo" in texto_lower or "modelos predictivos" in texto_lower
    }

    coincidencias = [k for k, v in criterios.items() if v]

    justificacion = f"""
    El candidato presenta un perfil de tipo {perfil} con nivel {nivel}.
    Su CV muestra coincidencias relevantes con la vacante en áreas como:
    {', '.join(coincidencias)}.
    Por esta razón, el sistema lo considera un perfil cercano a los requisitos
    técnicos y responsabilidades del puesto.
    """

    return justificacion.strip()

In [35]:
# Agregar justificación a los 5 mejores candidatos

df_top5["justificacion"] = df_top5.apply(
    lambda row: generar_justificacion(
        row["texto_extraido"],
        row["perfil"],
        row["nivel"]
    ),
    axis=1
)

df_top5[["ranking", "id", "nivel", "perfil", "score_similitud", "justificacion"]]

,ranking,id,nivel,perfil,score_similitud,justificacion
0,1,CV_11,Intermedio,Machine Learning Engineer,0.677967,El candidato presenta un perfil de tipo Machin...
1,2,CV_16,Senior,Machine Learning Engineer,0.668851,El candidato presenta un perfil de tipo Machin...
2,3,CV_07,Intermedio,Machine Learning Engineer,0.658535,El candidato presenta un perfil de tipo Machin...
3,4,CV_02,Junior,Machine Learning Engineer,0.635409,El candidato presenta un perfil de tipo Machin...
4,5,CV_09,Intermedio,Analista Business Intelligence,0.583487,El candidato presenta un perfil de tipo Analis...


### Interpretación de los 5 candidatos recomendados

A partir de la búsqueda semántica, el sistema recuperó los cinco candidatos con mayor similitud respecto a la vacante seleccionada. Los resultados muestran que los perfiles mejor posicionados corresponden principalmente a candidatos de tipo **Machine Learning Engineer**, lo cual es coherente con los requisitos técnicos de la vacante.

El candidato mejor evaluado fue **CV_11**, de nivel Intermedio y perfil Machine Learning Engineer, con un score de similitud aproximado de **0.6780**. Este perfil destaca por incluir experiencia en Python, Machine Learning, NLP, embeddings y modelos predictivos, competencias directamente relacionadas con la vacante.

En segundo lugar aparece **CV_16**, de nivel Senior y también perfil Machine Learning Engineer, con un score aproximado de **0.6689**. Este candidato cuenta con habilidades más avanzadas, como PySpark/Spark, MLOps, embeddings y modelos predictivos, por lo que representa un perfil sólido para el puesto.

El tercer lugar corresponde a **CV_07**, de nivel Intermedio y perfil Machine Learning Engineer, con un score aproximado de **0.6585**. Su experiencia en Python, Machine Learning, SQL y modelos predictivos lo convierte en un candidato alineado con las responsabilidades del puesto.

El cuarto lugar fue **CV_02**, de nivel Junior y perfil Machine Learning Engineer, con un score aproximado de **0.6354**. Aunque tiene menor experiencia profesional, el sistema lo identifica como relevante porque contiene habilidades técnicas cercanas a la vacante, como Python, Machine Learning, SQL, NLP y modelos predictivos.

Finalmente, el quinto lugar fue **CV_09**, de nivel Intermedio y perfil Analista Business Intelligence, con un score aproximado de **0.5835**. Aunque no es un perfil completamente especializado en Machine Learning, sí presenta competencias relacionadas con Python, SQL y análisis de datos, lo cual explica su aparición dentro de los cinco mejores candidatos.

En general, los resultados son coherentes con la lógica esperada del sistema, ya que los perfiles más cercanos a la vacante fueron priorizados por encima de perfiles menos relacionados, como desarrolladores de software o arquitectos de software. Esto indica que la búsqueda semántica logró identificar coincidencias relevantes entre la descripción del puesto y la información contenida en los CVs.

## 8. Reporte técnico de la arquitectura implementada

En esta actividad se diseñó un sistema inteligente multimodal para búsqueda de talento. La arquitectura implementada inicia con la generación de una base de conocimiento formada por 20 CVs sintéticos en formato PDF y PNG. Posteriormente, se extrae el texto de cada documento: para los archivos PDF se utiliza extracción directa de texto, mientras que para los archivos PNG se utiliza OCR.

Después de obtener el texto de cada CV, se construye una representación textual de la vacante seleccionada. Tanto los CVs como la vacante son convertidos en embeddings utilizando un modelo multilingüe de Sentence Transformers. Estos embeddings permiten representar el significado semántico de los documentos en forma vectorial.

Con los embeddings de los CVs se construye una base vectorial utilizando FAISS. Luego, el embedding de la vacante se compara contra la base vectorial para recuperar los candidatos con mayor similitud semántica. Finalmente, el sistema selecciona los 5 mejores candidatos y genera una justificación basada en la coincidencia entre sus habilidades, experiencia y los requisitos de la vacante.

La arquitectura general del sistema puede resumirse en las siguientes etapas:

1. Selección de vacante real.
2. Generación de CVs sintéticos.
3. Conversión de CVs a PDF y PNG.
4. Extracción de texto desde documentos multimodales.
5. Generación de embeddings.
6. Construcción de base vectorial.
7. Búsqueda semántica.
8. Ranking de candidatos.
9. Justificación de recomendaciones.

## Evaluación del comportamiento de la solución

El comportamiento del sistema fue adecuado, ya que los primeros lugares del ranking correspondieron principalmente a perfiles de Machine Learning Engineer, que son los más cercanos a la vacante seleccionada. Esto indica que la búsqueda semántica logró identificar coincidencias relevantes entre la descripción del puesto y el contenido de los CVs.

El candidato mejor posicionado fue CV_11, de nivel Intermedio y perfil Machine Learning Engineer, con un score aproximado de 0.6780. Este perfil incluye competencias como Python, Machine Learning, NLP, embeddings y modelos predictivos, por lo que resulta coherente que haya sido seleccionado como el candidato más cercano.

También fueron recomendados CV_16 y CV_07, ambos perfiles relacionados con Machine Learning Engineering. Esto refuerza la idea de que el sistema priorizó correctamente candidatos con habilidades técnicas alineadas a la vacante.

El quinto lugar fue ocupado por CV_09, un perfil de Analista Business Intelligence. Aunque no es el perfil más cercano a Machine Learning, sí cuenta con habilidades en Python, SQL y análisis de datos, lo que explica su aparición dentro del Top 5.

En general, el sistema mostró un comportamiento lógico y consistente. Los perfiles más cercanos a la vacante fueron priorizados sobre perfiles menos relacionados, como desarrolladores de software o arquitectos de software. Esto sugiere que el uso de embeddings y búsqueda vectorial es útil para apoyar procesos iniciales de búsqueda de talento.

## 9. Ideas adicionales para medir el desempeño de la solución

Además de la evaluación técnica realizada mediante el análisis del ranking obtenido, el desempeño de la solución podría medirse con otros métodos:

- **Evaluación por expertos:** un reclutador o especialista técnico podría revisar manualmente los candidatos recomendados y comparar si coinciden con su criterio profesional.

- **Precisión en el Top 5:** se podría medir cuántos de los cinco candidatos recomendados realmente cumplen con los requisitos principales de la vacante.

- **Comparación contra ranking manual:** se podría construir un ranking hecho por expertos y compararlo con el ranking generado por el sistema.

- **Recall de candidatos relevantes:** se podría analizar si el sistema logra recuperar la mayoría de los candidatos que sí son adecuados para la vacante.

- **Análisis de sesgos:** se podría revisar si el sistema favorece ciertos perfiles, niveles de experiencia o palabras clave de forma injustificada.

- **Pruebas con diferentes vacantes:** se podría evaluar el sistema con otras vacantes para comprobar si mantiene un comportamiento adecuado en distintos contextos.

- **Evaluación de robustez:** se podría probar el sistema con CVs en formatos distintos, con errores de OCR o con información incompleta para medir su estabilidad.

# **Conclusiones:**

Esta actividad permitió construir un sistema inteligente de búsqueda de talento utilizando técnicas de Procesamiento de Lenguaje Natural, embeddings y bases de datos vectoriales. El sistema fue capaz de analizar una vacante real, procesar una base de CVs sintéticos y recomendar los cinco candidatos con mayor similitud semántica.

Una parte importante de la actividad fue la generación de CVs sintéticos, ya que esto permitió simular un escenario real sin comprometer datos personales. Además, se evitó incluir información sensible como nombre, fotografía, género, edad, nacionalidad o estado civil, con el objetivo de reducir sesgos y proteger la privacidad.

El sistema mostró resultados coherentes, ya que priorizó perfiles relacionados con Machine Learning Engineering y Ciencia de Datos, que eran los más cercanos a la vacante seleccionada. Esto demuestra que la búsqueda semántica puede ser una herramienta útil para apoyar procesos de reclutamiento, especialmente cuando se tiene una gran cantidad de CVs.

Sin embargo, la solución no debe interpretarse como un reemplazo del reclutador humano. Su función principal es apoyar la selección inicial, reducir el tiempo de revisión y facilitar la identificación de perfiles potencialmente adecuados. La decisión final debe seguir considerando el juicio humano, entrevistas, experiencia real y criterios organizacionales.

En conclusión, el uso de NLP, embeddings y bases vectoriales permite construir sistemas inteligentes capaces de recuperar información relevante y generar recomendaciones justificadas. Este tipo de herramientas puede aportar valor en procesos de selección de talento, siempre que se utilicen de forma ética, transparente y complementaria al criterio humano.


# **Fin de la actividad : Búsqueda Inteligente de Talento**

In [36]:
# Comprimir la carpeta de CVs sintéticos en un archivo ZIP

import shutil
shutil.make_archive("cvs_sinteticos", "zip", "cvs_sinteticos")
print("Archivo generado: cvs_sinteticos.zip")

from google.colab import files
files.download("cvs_sinteticos.zip")

Archivo generado: cvs_sinteticos.zip


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>